### PR TITLE
Move the software API to a separate service `org.opensuse.DInstaller.Software`

### PR DESCRIPTION
**Pull request by imobachgs**
## Problem

This PRs moves the software management logic to a separate service, `org.opensuse.DInstaller.Software`. See #168 for further information.

## Solution

This is the first PR to move the software API to a separate `org.opensuse.DInstaller.Software`. It sets the basic infrastructure:

* Move the software API to a separate `SoftwareService`.
* Extend `bin/d-installer` to run all services at once or just one of them. More work is expected to improve this executable soon.
* Support for using a different `Y2DIR` per service.
* Adapt the UI to use the new software service.

## Testing

- [ ] *Added a new unit test*
- [x] *Tested manually*

## Tasks (other PRs)

- Keep track of the status
- API to add resolvables to the packages proposal (if needed)

### CODE REVIEW

File: service/bin/d-installer [hunk 1]
```diff
@@ -30,8 +30,42 @@ require "rubygems"
 require "bundler/setup"
 require "dinstaller/dbus/service_runner"
 
-service_name = $ARGV[0]&.to_sym || :manager
-service_runner = DInstaller::DBus::ServiceRunner.new(
-  service_name, logger: Logger.new($stdout)
-)
-service_runner.run
+# Run the service with the given name
```
Comment (joseivanlopez) [inline service/bin/d-installer:33]: NP: Runs

File: service/bin/d-installer [hunk 2]
```diff
@@ -30,8 +30,42 @@ require "rubygems"
...
+  logger = Logger.new($stdout, progname: name.to_s)
+  service_runner = DInstaller::DBus::ServiceRunner.new(
+    name, logger: logger
+  )
+  service_runner.run
+end
+
+ORDERED_SERVICES = [:software, :users, :manager].freeze
+
+# Start all the services
```
Comment (joseivanlopez) [inline service/bin/d-installer:47]: NP: Starts

File: service/lib/dinstaller/dbus/clients/software.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        end
+
+        # Starts the probing process
+        #
+        # If a block is given, the method returns inmmediatelly and the probing is performed in an
+        # asynchronous way.
+        #
+        # @param done [Proc] Block to execute once the probing is done
+        def probe(&done)
+          return dbus_object.Probe unless block_given?
```
Comment (joseivanlopez) [inline service/lib/dinstaller/dbus/clients/software.rb:64]: This line is not needed. You can simply use `dbus_object.Probe(&block)`, which will perform the D-Bus call without passing any block if the block was not given.

File: service/lib/dinstaller/dbus/clients/software.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+          dbus_object.Probe(&done)
+        end
+
+        # Perform package installation
```
Comment (joseivanlopez) [inline service/lib/dinstaller/dbus/clients/software.rb:69]: NP: in general, we tend to use third person.

**Review by dgdavid — APPROVED** [review]
So far, LGTM. Will you add unit testing in this or a subsequent PR?

File: service/lib/dinstaller/dbus/software.rb [hunk 5]
```diff
@@ -39,9 +40,13 @@ def initialize(backend, logger)
         @backend = backend
         @logger = logger
 
+        @progress_handler = DInstaller::Progress.new
```
Comment (joseivanlopez) [inline service/lib/dinstaller/dbus/software.rb:43]: We follow a different approach in `Manager`. The backend class `DInstaller::Manager` owns the progress object, and the D-Bus class configures the callbacks for such a progress object.

File: service/lib/dinstaller/dbus/software.rb [hunk 6]
```diff
@@ -64,7 +69,32 @@ def initialize(backend, logger)
...
+          backend.finish(progress_handler)
+        end
+
+        # Progress has struct with values:
+        #   s message
+        #   t total major steps to do
+        #   t current major step (0-based)
+        #   t total minor steps. Can be zero which means no minor steps
+        #   t current minor step
+        dbus_reader :progress, "(stttt)"
```
Comment (joseivanlopez) [inline service/lib/dinstaller/dbus/software.rb:95]: Maybe we should start ignoring the minor steps. I think we will drop minor steps sooner than later.

File: service/lib/dinstaller/dbus/software_service.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      # @param config [Config] Configuration object
+      # @param logger [Logger]
+      def initialize(config, logger = nil)
+        @backend = DInstaller::Software.new(config, logger)
+        @logger = logger || Logger.new($stdout)
+        @bus = ::DBus::SystemBus.instance
+      end
+
+      # Exports the software object through the D-Bus service
+      def export
```
Comment (joseivanlopez) [inline service/lib/dinstaller/dbus/software_service.rb:50]: At some point we will need a `DBus::Base` class.

File: service/lib/dinstaller/dbus/software.rb (hunk 6 above)
Comment (imobachgs) [reply service/lib/dinstaller/dbus/software.rb:95]: I agree. But I preferred to keep them consistent until we have a final working version. Perhaps we find those bits useful at the end.

**Review by joseivanlopez — APPROVED** [review]
LGTM

**Pull request merged**